File: JavaPort/src/view/TextureRepository.java
package view;

import java.io.*;
import java.nio.ByteBuffer;

import data.TileType;
import de.matthiasmann.twl.utils.PNGDecoder;
import de.matthiasmann.twl.utils.PNGDecoder.Format;
import static org.lwjgl.opengl.GL11.*;

public class TextureRepository {
	private static int nextTextureNumber = 1;

	private static Texture mountianShort;
	private static Texture mountianTall;
	private static Texture forest;
	private static Texture grass;

	private static Texture waterCenter;
	private static Texture waterLowerLeft;
	private static Texture waterLowerMiddle;
	private static Texture waterLowerRight;
	private static Texture waterMiddleLeft;
	private static Texture waterMiddleRight;
	private static Texture waterUpperLeft;
	private static Texture waterUpperMiddle;
	private static Texture waterUpperRight;
	private static Texture waterLowerChannelEnd;
	private static Texture waterUpperChannelEnd;
	private static Texture waterRightChannelEnd;
	private static Texture waterLeftChannelEnd;

	private static Texture waterChannelHorizontal;
	private static Texture waterChannelVertical;

	private static Texture waterIsland;
	private static Texture waterUpperRightTip;
	private static Texture waterUpperLeftTip;
	private static Texture waterLowerRightTip;
	private static Texture waterLowerLeftTip;

	public static void initalize() {
		grass = bindTexture("Art/Tiles/AW/AWgrass.png");
		mountianShort = bindTexture("Art/Tiles/AW/AWmountianShort.png");
		mountianTall = bindTexture("Art/Tiles/AW/AWmountianTall.png");
		forest = bindTexture("Art/Tiles/AW/AWforest.png");

		waterIsland = bindTexture("Art/Tiles/AW/Water/AWWaterIsland.png");
		waterCenter = bindTexture("Art/Tiles/AW/Water/AWWaterCenter.png");
		waterLowerLeft = bindTexture("Art/Tiles/AW/Water/AWWaterLowerLeft.png");
		waterLowerMiddle = bindTexture("Art/Tiles/AW/Water/AWWaterLowerMiddle.png");
		waterLowerRight = bindTexture("Art/Tiles/AW/Water/AWWaterLowerRight.png");
		waterMiddleLeft = bindTexture("Art/Tiles/AW/Water/AWWaterMiddleLeft.png");
		waterMiddleRight = bindTexture("Art/Tiles/AW/Water/AWWaterMiddleRight.png");
		waterUpperLeft = bindTexture("Art/Tiles/AW/Water/AWWaterUpperLeft.png");
		waterUpperMiddle = bindTexture("Art/Tiles/AW/Water/AWWaterUpperMiddle.png");
		waterUpperRight = bindTexture("Art/Tiles/AW/Water/AWWaterUpperRight.png");

		waterLowerChannelEnd = bindTexture("Art/Tiles/AW/Water/AWWaterLowerChannelEnd.png");
		waterUpperChannelEnd = bindTexture("Art/Tiles/AW/Water/AWWaterUpperChannelEnd.png");
		waterRightChannelEnd = bindTexture("Art/Tiles/AW/Water/AWWaterRightChannelEnd.png");
		waterLeftChannelEnd = bindTexture("Art/Tiles/AW/Water/AWWaterLeftChannelEnd.png");

		waterChannelVertical = bindTexture("Art/Tiles/AW/Water/AWWaterChannelVertical.png");
		waterChannelHorizontal = bindTexture("Art/Tiles/AW/Water/AWWaterChannelHorizontal.png");

		waterUpperRightTip = bindTexture("Art/Tiles/AW/Water/AWWaterUpperRightTip.png");
		;
		waterUpperLeftTip = bindTexture("Art/Tiles/AW/Water/AWWaterUpperLeftTip.png");
		;
		waterLowerRightTip = bindTexture("Art/Tiles/AW/Water/AWWaterLowerRightTip.png");
		;
		waterLowerLeftTip = bindTexture("Art/Tiles/AW/Water/AWWaterLowerLeftTip.png");
	}

	public static Texture GetTileTexture(TileType tile) {
		switch (tile) {
		case Grass:
			return grass;
		case MountainShort:
			return mountianShort;
		case MountainTall:
			return mountianTall;
		case Forest:
			return forest;
		case Water:
			return waterCenter;
		default:
			return new Texture(0, 0, 0);
		}
	}

	public static Texture GetWater(WaterEdgeType edgeType) {
		switch (edgeType) {
		case LowerLeft:
			return waterLowerLeft;
		case LowerMiddle:
			return waterLowerMiddle;
		case LowerRight:
			return waterLowerRight;
		case MiddleLeft:
			return waterMiddleLeft;
		case MiddleRight:
			return waterMiddleRight;
		case UpperLeft:
			return waterUpperLeft;
		case UpperMiddle:
			return waterUpperMiddle;
		case UpperRight:
			return waterUpperRight;
		case Center:
			return waterCenter;
		case LeftChannelEnd:
			return waterLeftChannelEnd;
		case RightChannelEnd:
			return waterRightChannelEnd;
		case UpperChannelEnd:
			return waterUpperChannelEnd;
		case LowerChannelEnd:
			return waterLowerChannelEnd;
		case ChannelHorizontal:
			return waterChannelHorizontal;
		case ChannelVertical:
			return waterChannelVertical;
		case WaterIsland:
			return waterIsland;

		case LowerLeftTip:
			return waterLowerLeftTip;
		case LowerRightTip:
			return waterLowerRightTip;
		case UpperLeftTip:
			return waterUpperLeftTip;
		case UpperRightTip:
			return waterUpperRightTip;

		default:
			return new Texture(0, 0, 0);
		}
	}

	private static Texture bindTexture(String fileName) {

		int currentTextureNumber = nextTextureNumber;
		nextTextureNumber++;
		try {
			InputStream in = new FileInputStream(fileName);
			PNGDecoder decoder = new PNGDecoder(in);

			int imageDataSize = 4 * decoder.getWidth() * decoder.getHeight();
			ByteBuffer imageDataBuffer = ByteBuffer.allocateDirect(imageDataSize);
			decoder.decode(imageDataBuffer, 4 * decoder.getWidth(), Format.RGBA);
			imageDataBuffer.flip();
			in.close();

			glBindTexture(GL_TEXTURE_2D, currentTextureNumber);
			glPixelStorei(GL_UNPACK_ALIGNMENT, 1);

			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);

			glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_MODULATE);

			glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, decoder.getWidth(), decoder.getHeight(), 0, GL_RGBA,
					GL_UNSIGNED_BYTE, imageDataBuffer);

			return new Texture(currentTextureNumber, decoder.getWidth(), decoder.getHeight());

		} catch (Exception ex) {
			System.out.println("Error Loading texture named " + fileName);
			ex.printStackTrace();

			return new Texture(0, 0, 0);
		}

	}
}

File: XnaVersion/BuildingPrototype/BuildingPrototype/View/TextureRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Content;
using MapEditor.Data;
using Microsoft.Xna.Framework;

namespace MapEditor.View
{
    static class TextureRepository
    {
        private static Texture2D mountianShort;
        private static Texture2D mountianTall;
        private static Texture2D forest;
        private static Texture2D grass;
        private static Texture2D floor;
        private static Texture2D wall;
        private static Texture2D door;

        private static Texture2D waterCenter;
        private static Texture2D waterLowerLeft;
        private static Texture2D waterLowerMiddle;
        private static Texture2D waterLowerRight;
        private static Texture2D waterMiddleLeft;
        private static Texture2D waterMiddleRight;
        private static Texture2D waterUpperLeft;
        private static Texture2D waterUpperMiddle;
        private static Texture2D waterUpperRight;
        private static Texture2D waterLowerChannelEnd;
        private static Texture2D waterUpperChannelEnd;
        private static Texture2D waterRightChannelEnd;
        private static Texture2D waterLeftChannelEnd;

        private static Texture2D waterChannelHorizontal;
        private static Texture2D waterChannelVertical;

        private static Texture2D waterIsland;
        private static Texture2D waterUpperRightTip;
        private static Texture2D waterUpperLeftTip;
        private static Texture2D waterLowerRightTip;
        private static Texture2D waterLowerLeftTip;

        private static Texture2D soilder;
        private static Texture2D pixel;

        // TODO: Clean up the water texture initalizing
        public static void Initalize(ContentManager content, GraphicsDevice graphics)
        {
            grass = content.Load<Texture2D>(@"art/tiles/AWgrass");
            mountianShort = content.Load<Texture2D>(@"art/tiles/AWmountianShort");
            mountianTall = content.Load<Texture2D>(@"art/tiles/AWmountianTall");
            forest = content.Load<Texture2D>(@"art/tiles/AWforest");
            floor = content.Load<Texture2D>(@"art/tiles/Floor");
            wall = content.Load<Texture2D>(@"art/tiles/Wall");
            door = content.Load<Texture2D>(@"art/tiles/Door");

            waterIsland = content.Load<Texture2D>(@"art/tiles/AWWaterIsland");
            waterCenter = content.Load<Texture2D>(@"art/tiles/AWWaterCenter");
            waterLowerLeft = content.Load<Texture2D>(@"art/tiles/AWWaterLowerLeft");
            waterLowerMiddle = content.Load<Texture2D>(@"art/tiles/AWWaterLowerMiddle");
            waterLowerRight = content.Load<Texture2D>(@"art/tiles/AWWaterLowerRight");
            waterMiddleLeft = content.Load<Texture2D>(@"art/tiles/AWWaterMiddleLeft");
            waterMiddleRight = content.Load<Texture2D>(@"art/tiles/AWWaterMiddleRight");
            waterUpperLeft = content.Load<Texture2D>(@"art/tiles/AWWaterUpperLeft");
            waterUpperMiddle = content.Load<Texture2D>(@"art/tiles/AWWaterUpperMiddle");
            waterUpperRight = content.Load<Texture2D>(@"art/tiles/AWWaterUpperRight");

            waterLowerChannelEnd = content.Load<Texture2D>(@"art/tiles/AWWaterLowerChannelEnd");
            waterUpperChannelEnd = content.Load<Texture2D>(@"art/tiles/AWWaterUpperChannelEnd");
            waterRightChannelEnd = content.Load<Texture2D>(@"art/tiles/AWWaterRightChannelEnd");
            waterLeftChannelEnd = content.Load<Texture2D>(@"art/tiles/AWWaterLeftChannelEnd");

            waterChannelVertical = content.Load<Texture2D>(@"art/tiles/AWWaterChannelVertical");
            waterChannelHorizontal = content.Load<Texture2D>(@"art/tiles/AWWaterChannelHorizontal");

            waterUpperRightTip = content.Load<Texture2D>(@"art/tiles/AWWaterUpperRightTip"); ;
            waterUpperLeftTip = content.Load<Texture2D>(@"art/tiles/AWWaterUpperLeftTip"); ;
            waterLowerRightTip = content.Load<Texture2D>(@"art/tiles/AWWaterLowerRightTip"); ;
            waterLowerLeftTip = content.Load<Texture2D>(@"art/tiles/AWWaterLowerLeftTip");

            soilder = content.Load<Texture2D>(@"art/actors/Soldier");

            pixel = new Texture2D(graphics, 1, 1);
            Color[] whitePixel = {Color.White};
            pixel.SetData<Color>(whitePixel);


            GlobalSettings.TileSize = grass.Width;
        }
        public static Texture2D GetActorTexture(ActorType type)
        {
            switch(type)
            {
                case ActorType.SoldierOne:
                    return soilder;
                default:
                    throw new ArgumentException(); // TODO: make a more informative custom exception to throw
            }
        }
        public static Texture2D GetTileTexture(TileType tile)
        {
            switch(tile)
            {
                case TileType.Grass:
                    return grass;
                case TileType.MountainShort:
                    return mountianShort;
                case TileType.MountainTall:
                    return mountianTall;
                case TileType.Forest:
                    return forest;
                case TileType.Water:
                    return waterCenter;
                case TileType.Floor:
                    return floor;
                case TileType.Door:
                    return door;
                case TileType.Wall:
                    return wall;
                default:
                    throw new ArgumentException();
            }
        }
        public static Texture2D GetWater(WaterEdgeType edgeType)
        {
            switch(edgeType)
            {
                case WaterEdgeType.LowerLeft:
                    return waterLowerLeft;
                case WaterEdgeType.LowerMiddle:
                    return waterLowerMiddle;
                case WaterEdgeType.LowerRight:
                    return waterLowerRight;
                case WaterEdgeType.MiddleLeft:
                    return waterMiddleLeft;
                case WaterEdgeType.MiddleRight:
                    return waterMiddleRight;
                case WaterEdgeType.UpperLeft:
                    return waterUpperLeft;
                case WaterEdgeType.UpperMiddle:
                    return waterUpperMiddle;
                case WaterEdgeType.UpperRight:
                    return waterUpperRight;
                case WaterEdgeType.Center:
                    return waterCenter;
                case WaterEdgeType.LeftChannelEnd:
                    return waterLeftChannelEnd;
                case WaterEdgeType.RightChannelEnd:
                    return waterRightChannelEnd;
                case WaterEdgeType.UpperChannelEnd:
                    return waterUpperChannelEnd;
                case WaterEdgeType.LowerChannelEnd:
                    return waterLowerChannelEnd;
                case WaterEdgeType.ChannelHorizontal:
                    return waterChannelHorizontal;
                case WaterEdgeType.ChannelVertical:
                    return waterChannelVertical;
                case WaterEdgeType.WaterIsland:
                    return waterIsland;

                case WaterEdgeType.LowerLeftTip:
                    return waterLowerLeftTip;
                case WaterEdgeType.LowerRightTip:
                    return waterLowerRightTip;
                case WaterEdgeType.UpperLeftTip:
                    return waterUpperLeftTip;
                case WaterEdgeType.UpperRightTip:
                    return waterUpperRightTip;

                default:
                    throw new ArgumentException("Invalid Water Type");
            }
        }

        public static Texture2D Pixel
        {
            get { return pixel; }
        }
    }
}
